src/main/java/com/github/shyiko/mysql/binlog/json/JsonBinary.java

```diff
@@ -0,0 +1,1012 @@
+                    "' in first byte of a JSON opaque value");
+        }
+        // Read the data length ...
+        int length = readVariableInt();
+
+        switch (type) {
+            case DECIMAL:
+            case NEWDECIMAL:
+                // See 'Json_decimal::convert_from_binary'
+                // https://github.com/mysql/mysql-server/blob/5.7/sql/json_dom.cc#L1625
+                parseDecimal(length, formatter);
+                break;
+
+            // All dates and times are in one of these types
+            // See 'Json_datetime::to_packed' for details
+            // https://github.com/mysql/mysql-server/blob/5.7/sql/json_dom.cc#L1681
+            // which calls 'TIME_to_longlong_packed'
+            // https://github.com/mysql/mysql-server/blob/5.7/sql-common/my_time.c#L2005
+            case DATE:
+                parseDate(formatter);
+                break;
+            case TIME:
+            case TIME_V2:
+                parseTime(formatter);
+                break;
+            case DATETIME:
+            case DATETIME_V2:
+                parseDatetime(formatter);
+                break;
+            default:
+                parseOpaqueValue(type, length, formatter);
+        }
+    }
+
+    /**
+     * Parse a {@code DATE} value, which is stored using the same format as {@code DATETIME}:
+     * 5 bytes + fractional-seconds storage. However, the hour, minute, second, and fractional seconds are ignored.
+     * <p>
+     * The non-fractional part is 40 bits:
+     *
+     * <pre>
+     *  1 bit  sign           (1= non-negative, 0= negative)
+     *  17 bits year*13+month  (year 0-9999, month 0-12)
+     *   5 bits day            (0-31)
+     *   5 bits hour           (0-23)
+     *   6 bits minute         (0-59)
+     *   6 bits second         (0-59)
+     * </pre>
+     *
+     * The fractional part is typically dependent upon the <i>fsp</i> (i.e., fractional seconds part) defined by
+     * a column, but in the case of JSON it is always 3 bytes.
+     * <p>
+     * The format of all temporal values is outlined in the <a href=
+     * "https://dev.mysql.com/doc/internals/en/date-and-time-data-type-representation.html">MySQL documentation</a>,
+     * although since the MySQL {@code JSON} type is only available in 5.7, only version 2 of the date-time formats
+     * are necessary.
+     *
+     * @param formatter the formatter to be notified of the parsed value; may not be null
+     * @throws IOException if there is a problem reading the JSON value
+     */
+    protected void parseDate(JsonFormatter formatter) throws IOException {
+        long raw = readInt64();
+        long value = raw >> 24;
+        int yearMonth = (int) (value >> 22) % (1 << 17); // 17 bits starting at 22nd
+        int year = yearMonth / 13;
+        int month = yearMonth % 13;
+        int day = (int) (value >> 17) % (1 << 5); // 5 bits starting at 17th
+        formatter.valueDate(year, month, day);
+    }
```

Comment:
Note that this and the other methods to parse temporal values are indeed slightly different from similar methods in `AbstractRowsEventDataDeserializer`. In particular, these are in little-endian form, which is contrary to the documentation and `AbstractRowsEventDataDeserializer` methods. However, this was verified through testing and via the MySQL source code.